File: quotation-service/src/routes/api-keys.ts
import { Router } from 'express';
import { listApiKeys, createApiKey, revokeApiKey } from '../db.js';
import { requireAuth, requireRole } from '../auth/middleware.js';

export const apiKeysRouter = Router();

apiKeysRouter.use(requireAuth, requireRole('admin'));

apiKeysRouter.get('/', (_req, res) => {
  return res.json({ success: true, apiKeys: listApiKeys() });
});

apiKeysRouter.post('/', (req, res) => {
  const { label } = req.body || {};
  if (!label || String(label).trim() === '') {
    return res.status(400).json({ success: false, error: 'label est requis.' });
  }

  const { record, plainToken } = createApiKey(String(label).trim(), req.user!.userId);
  // The plaintext token is only ever returned here, at creation time — it is
  // not recoverable afterwards (only the hash is stored).
  return res.status(201).json({ success: true, apiKey: record, token: plainToken });
});

apiKeysRouter.delete('/:id', (req, res) => {
  const id = Number(req.params.id);
  revokeApiKey(id);
  return res.status(204).send();
});

File: inference-service/app/routers/api_keys.py
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel

from .. import db
from ..auth.dependencies import require_role

router = APIRouter(prefix="/api/api-keys", tags=["api-keys"], dependencies=[Depends(require_role("admin"))])


class CreateApiKeyRequest(BaseModel):
    label: str


@router.get("")
def list_api_keys():
    return {"success": True, "apiKeys": db.list_api_keys()}


@router.post("", status_code=status.HTTP_201_CREATED)
def create_api_key(payload: CreateApiKeyRequest, current: dict = Depends(require_role("admin"))):
    if not payload.label.strip():
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "label est requis.")
    record, plain_token = db.create_api_key(payload.label.strip(), current["userId"])
    # The plaintext token is only ever returned here, at creation time.
    return {"success": True, "apiKey": record, "token": plain_token}


@router.delete("/{key_id}", status_code=status.HTTP_204_NO_CONTENT)
def revoke_api_key(key_id: int):
    db.revoke_api_key(key_id)
